oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,438 @@
+should always be to eliminate instructions entirely by making everything
+self-explanatory, or as close to it as possible.” This principal should
+also be applied to data modeling. Data models should have descriptive
+names clearly identifying the data that the model holds. The field names
+should make sense to most people familiar with the domain, without
+needing much clarifying documentation.
+
+Prioritize ease of analysis
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For every hour of engineer-time spent creating a data model, several
+thousand more are spent using the model and analyzing its data. Taking
+the time to consider how the data will be used, and thoughtfully design
+our data representation, can potentially save many hours during the
+analysis phase.
+
+This is especially true since changing data models, once they are in
+use, can be time-consuming. A data model change in a core application
+can affect many other systems. Small changes to a data model could
+possibly cause hours or days of work for different teams throughout the
+organization:
+
+-  Data engineering and Analytics to update their workflows
+
+-  BI teams to update reports that utilize this data,
+
+-  Devops to deploy migrations to large tables, with the risk of
+      potential downtime.
+
+Due to the potentially high cost of changing a model, it pays to get it
+as correct as possible the first time.
+
+Think about the person who will analyze this data later and the person
+who manages the system day to day. (That person could possibly be future
+you!) Try to do as much careful design up front to make your quality of
+life better later.
+
+History is best (but events are a close second)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Accuracy is best when history about a change is captured at the database
+level when the change happens. Changes in data relating to business
+continuity should be stored historically where the change is made,
+preferably in the same system as the system of record. This allows us to
+reconstruct the data at different points with much greater reliability.
+
+If, for some reason, it is not possible to create history at the time of
+entry (for example, if we expect the table to become too large or the
+writes to be too frequent), We recommend that an event be issued in its
+place.
+
+Fire events for CUD operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We highly recommend firing events for any create, update or delete (CUD)
+operations that happen within the OpenedX ecosystem. The cost of firing
+and storing these events is low, and will allow teams in the future to
+reconstruct the state of the application at various times.
+
+Data Modeling Standards
+-----------------------
+
+All models in the edX ecosystem should have:
+
+-  A Primary Key
+
+   -  Should be named: “id”
+
+   -  Use numeric based identifiers
+
+   -  A key size large enough so the keys will not run out for a very
```

Comment:
For those of us not experienced with defining custom `id` fields, is the default Django `id` generally sufficient?

Reply:
+1 -- we should make sure it's explicitly called out. The default `id` gives you a signed 4-byte int, so the max is (2^31 - 1). Sufficient for most of our tables, but definitely not all.

Reply:
Might want to explicitly call out `UnsignedBigIntAutoField` for Django models here if that's what we want to be used.

Reply:
If we go with this change adding a code block with a sample model will make it so that more people can easily do the right thnig.

Reply:
I'm not strongly advocating either way, but keep in mind that for InnoDB, the primary key is copied into every index on the table.

Reply:
Sure space is cheap, but memory always isn't. These rows will eventually be loaded in to memory for analysis and joined on. 


I don't think that blanket usage of UnsignedBigIntAutoField is the right thing here. I think that those designing the system need to think about what the correct choice is for the application they are designing.

Reply:
Ended up going with BigAutoField as the recommended field type.